tests/test_glogger.py

```diff
@@ -0,0 +1,233 @@
+        # Check that one message was dropped, and an additional warning message was added
+        assert_buffer_attributes(handler, count=3, head_serial_no=1, next_serial_no=4)
+        last_message = json.loads(handler.messages_buffer.buffer[-1])
+        assert last_message["severity"] == logging.WARNING
+        assert last_message["message"] == "Maximum total length (4000) exceeded. Dropped 1 messages."
+
+
+def test_content_type_json():
+    """Test handler sends valid JSON."""
+
+    class ReqHandler(BaseHTTPRequestHandler):
+        def do_POST(self):
+            assert self.headers["Content-Type"] == "application/json"
+            json_data = json.load(self.rfile)
+            assert isinstance(json_data, dict)
+            assert set(json_data.keys()) == {"batch_id", "metadata", "logs"}
+            logs = json_data["logs"]
+            assert isinstance(logs, list)
+            for log_item in logs:
+                assert isinstance(log_item, dict)
+                assert set(log_item.keys()) == {"serial_no", "level", "timestamp", "logger_name", "message"}
+            self.send_response(200, "OK")
+            self.end_headers()
+
+    with ExitStack() as exit_stack:
+        logs_server = LogsServer(("localhost", 0), ReqHandler)
+        exit_stack.callback(logs_server.server_close)
+
+        handler = HttpBatchRequestsHandler(logs_server.authority, max_total_length=10000)
+        exit_stack.callback(handler.stop)
+
+        logging.basicConfig(force=True, handlers=[handler], level=0)
+        logging.info("A" * 1000)
+        logging.info("B" * 2000)
+        logging.info("C" * 3000)
+        logging.info("D" * 4000)
+        logs_server.handle_request()
+        assert logs_server.processed > 0
+
+
+def test_error_flushing():
+    """Test handler logs a message when it get an error response from server."""
+
+    class ErrorRequestHandler(BaseHTTPRequestHandler):
+        def do_POST(self):
+            self.send_response(403, "Forbidden")
+            self.end_headers()
+
+    with ExitStack() as exit_stack:
+        logs_server = LogsServer(("localhost", 0), ErrorRequestHandler)
+        exit_stack.callback(logs_server.server_close)
+
+        handler = HttpBatchRequestsHandler(logs_server.authority, max_total_length=10000)
+        exit_stack.callback(handler.stop)
+
+        logging.basicConfig(force=True, handlers=[handler], level=0)
+        logging.info("A" * 1000)
+        logging.info("B" * 2000)
+        logging.info("C" * 3000)
+        logging.info("D" * 4000)
+        logs_server.handle_request()
+        assert logs_server.processed > 0
+        last_message = json.loads(handler.messages_buffer.buffer[-1])
+        assert last_message["severity"] == logging.ERROR
+        assert last_message["message"] == "Error posting to server"
+
+
+def test_truncate_long_message():
+    """Test message is truncated and marked accordingly if it's longer than max message size."""
+    with ExitStack() as exit_stack:
```

Comment:
The exitStack appears to many time - I suggest to use pytest fixtures to wrap this up more nicely